Wrapper: Allow SHA-256 verification of the downloaded gradle version

dschuermann opened the pull request:
Allow SHA-256 verification of the downloaded gradle version. This is similar to what can be done via the gradle-witness plugin for Maven dependencies [0] but for the gradle wrapper itself. To use this functionality do a sha256sum gradle-2.3-bin.zip and add the resulting hash to gradle-wrapper.properties, e.g., distributionSha256Sum=010dd9f31849abc3d5644e282943b1c1c355f8e2635c5789833979ce590a3774

I explicitly put this code into public domain to avoid signing the CLA.

[0] https://github.com/WhisperSystems/gradle-witness

dschuermann commented:
OK, wait with merging, I will refactor some things

dschuermann commented:
Done with re-factoring.
I am open for reviews and feedback!

alkemist commented:
@hansd do you know what the legal implication here is of @dschuermann explicitly declaring that this is public domain and therefore no CLA being required? Does this meet our legal obligations?

@dschuermann thanks for the contribution. It may take a little bit of time to resolve this CLA issue. The implementation looks good to me, but we'll take a deeper look once the CLA issue is resolved.

hansd commented:
On Thu, Apr 30, 2015 at 7:49 PM, Luke Daley <notifications@github.com>
wrote:

> @hansd <https://github.com/hansd> do you know what the legal implication
> here is of @dschuermann <https://github.com/dschuermann> explicitly
> declaring that this is public domain and therefore no CLA being required?
> Does this meet our legal obligations?
>

We need a CLA to protect our users. We follow the Apache recommendations
and more or less use their CLA.

Hans

> @dschuermann <https://github.com/dschuermann> thanks for the
> contribution. It may take a little bit of time to resolve this CLA issue.
> The implementation looks good to me, but we'll take a deeper look once the
> CLA issue is resolved.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gradle/gradle/pull/448#issuecomment-98030430>.
>

dschuermann commented:
I hoped to get around signing the CLA, but anyway, it's now signed.

alkemist commented:
Thanks for signing the CLA @dschuermann. This makes things much simpler for us.

alkemist commented on subprojects/wrapper/src/main/java/org/gradle/wrapper/Install.java:
```diff
@@ -77,7 +82,48 @@ public File call() throws Exception {
+            throws Exception {
+        // check SHA-256 checksum verification
+        if (distributionSha256Sum != null) {
+            logger.log("Verifying " + localZipFile.getName());
```
This could be a touch more descriptive.

alkemist commented on subprojects/wrapper/src/main/java/org/gradle/wrapper/Install.java:
```diff
@@ -77,7 +82,48 @@ public File call() throws Exception {
+            logger.log("Verifying " + localZipFile.getName());
+
+            if (!distributionSha256Sum.equals(calculateSha256Sum(localZipFile))) {
+                deleteDir(distDir);
```
This makes me slightly uncomfortable. Might be better to just throw without deleting, and include this path in the error message so the user can delete if they want to.

alkemist commented:
There's two things missing for this:

1. Some test coverage (an integration test each for failure/success in https://github.com/gradle/gradle/blob/master/subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperConcurrentDownloadTest.groovy#L33-33 would do the trick)
2. Some documentation on this capability (added to this file: https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/userguide/gradleWrapper.xml#L17-17) 

@dschuermann up for any/all of these?

dschuermann commented:
@alkemist I added a small doc, removed the deletion of the cached gradle distribution and improved the output when verification fails.
Unfortunately, due to time constraints I am currently unable to provide a test case.

dschuermann commented:
FYI: I received the emails that I successfully signed the CLA.

More information needed (besides the test cases)?

alkemist commented:
@dschuermann it's just waiting on our side. We've got a long backlog of pull requests and little discretionary time at the moment to process them. We are internally improving our processes to allow us to deal with this better.

dschuermann commented:
@alkemist great! Thanks for reviewing

eighthave commented:
I think this is a great improvement over the current state of affairs. What about including some sha256 sums of well known release files in the gradle-wrapper by default?  Like hard-coded, so people automatically gain this protection, instead of having thousands of people have to manually add it.

mark-vieira commented:
@dschuermann We'd like to go ahead a merge this for inclusion in Gradle 2.6. The main missing piece is test coverage. Could you please add some tests as per @alkemist's comment above?

dschuermann commented:
@mark-vieira Unfortunately, I really don't have more free time to do it, sry.

mark-vieira commented:
@dschuermann No worries, thanks for the contribution. We'll take it from here.

ffgiff commented:
This functionality is currently broken in IDEs such as Android Studio, see #9361. @dschuermann would you be able to fix this?
Thanks.

dschuermann commented:
@ffgiff Unfortunately, I don't have time for this right now.

eighthave commented:
besides, this has got to be a bug in Android Studio, as far as I've seen

ffgiff commented:
> besides, this has got to be a bug in Android Studio, as far as I've seen

My reading of #9361 is that something is missing from a Gradle API to let tools know about verification failures, and AS works around this by telling users to stop verifying the download. If other IDEs have some other way to detect failures and are not affected, then sure, AS could just find a better workaround.

carstenhag commented:
Bug being tracked @ #9361
